### PR TITLE
boards: teensy4: Add ethernet functionality

### PR DESCRIPTION
**Pull request by bdkrae**
- Adds support for Texas Instruments **DP83825** RMII 10/100 ethernet transceiver. Driver supports:
        - optional RESET pin
        - optional INT pin
        - config option for 25 or 50MHz RMII clock
- Updates teensy board configs for use with new `eth_nxp_enet.c` driver

### CODE REVIEW

**Review by decsny — DISMISSED** [review]

**Review by vaishnavachath — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Comment by bdkrae** [comment]
> Could you add a device entry to `tests/drivers/build_all/ethernet/`?

Yes, but would prefer separate PR as there are some more update tasks regarding tests for this board.

**Comment by bdkrae** [comment]
@jukkar could you have a look at this PR?

File: drivers/ethernet/phy/phy_ti_dp83825.c [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+		goto done;
+	}
+
+	/* Setup advertising register */
+	if (speeds & LINK_FULL_100BASE_T) {
+		anar |= MII_ADVERTISE_100_FULL;
+	} else {
+		anar &= ~MII_ADVERTISE_100_FULL;
+	}
+	if (speeds & LINK_HALF_100BASE_T) {
```
Comment (jukkar) [inline drivers/ethernet/phy/phy_ti_dp83825.c:411]: Please add empty line after `if {}` block here and in whole file.

File: drivers/ethernet/phy/phy_ti_dp83825.c [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+	if (ret) {
+		LOG_ERR("Error reading phy (%d) link partner register", config->addr);
+		k_mutex_unlock(&data->mutex);
+		return ret;
+	}
+
+	/* Unlock mutex */
+	k_mutex_unlock(&data->mutex);
+
+	uint32_t mutual_capabilities = anar & anlpar;
```
[the remark this answers is not in the document]
Comment (bdkrae) [reply drivers/ethernet/phy/phy_ti_dp83825.c]: fixed

File: drivers/ethernet/phy/phy_ti_dp83825.c [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+	} else if (mutual_capabilities & MII_ADVERTISE_10_FULL) {
+		state->speed = LINK_FULL_10BASE_T;
+	} else if (mutual_capabilities & MII_ADVERTISE_10_HALF) {
+		state->speed = LINK_HALF_10BASE_T;
+	} else {
+		ret = -EIO;
+	}
+
+result:
+	if (memcmp(&old_state, state, sizeof(struct phy_link_state)) != 0) {
```
[the remark this answers is not in the document]
Comment (bdkrae) [reply drivers/ethernet/phy/phy_ti_dp83825.c:228]: no, changed line 223 to `return -EIO;`

File: drivers/ethernet/phy/phy_ti_dp83825.c [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+	}
+
+	/* Enable link state changed interrupt*/
+	reg_val |= PHY_TI_DP83825_MISR_REG_LSCE;
+
+	/* Write settings to misr register */
+	ret = phy_ti_dp83825_write(dev, PHY_TI_DP83825_MISR_REG, reg_val);
+	if (ret) {
+		return ret;
+	}
```
[the remark this answers is not in the document]
Comment (bdkrae) [reply drivers/ethernet/phy/phy_ti_dp83825.c]: done

File: drivers/ethernet/phy/phy_ti_dp83825.c (hunk 1 above)
Comment (bdkrae) [reply drivers/ethernet/phy/phy_ti_dp83825.c:410]: done

**Comment by pdgendt** [comment]
> > Could you add a device entry to `tests/drivers/build_all/ethernet/`?
> 
> Yes, but would prefer separate PR as there are some more update tasks regarding tests for this board.

It should be a simple entry to make sure the driver is built in CI for `ti,dp83825` PHY devices.

**Comment by bdkrae** [comment]
> > > Could you add a device entry to `tests/drivers/build_all/ethernet/`?
> > 
> > 
> > Yes, but would prefer separate PR as there are some more update tasks regarding tests for this board.
> 
> It should be a simple entry to make sure the driver is built in CI for `ti,dp83825` PHY devices.

I´m not sure if I understand correctly. Do I need to change `boards/pjrc/teensy4/teensy41.yml` to include entry `netif:eth`?

**Comment by pdgendt** [comment]
> I´m not sure if I understand correctly. Do I need to change `boards/pjrc/teensy4/teensy41.yml` to include entry `netif:eth`?

I think that should be sufficient, you can do a (dry) run using twister to see if the board is included.

**Comment by pdgendt** [comment]
For some reason, the boards were not being built by default, I've opened #79537 to fix the CI coverage, the teensy41 board should be added too.

**Comment by pdgendt** [comment]
@bdkrae please rebase on latest `main`, the CI issues have been fixed.

**Comment by bdkrae** [comment]
added `netif:eth`, works as expected
image

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by bdkrae** [comment]
@pdgendt Who can merge this? Or do we need more approvals?

**Comment by pdgendt** [comment]
Assignee approval is required. cc @vaishnavachath

**Review by vaishnavachath — APPROVED** [review]

**Pull request merged**